File: Dart/Arrays&Hashing/Problems/5-MaximumSubarray(Kadane’s Algorithm)/MaximumSubarray(Kadane’sAlgorithm)Pro.dart
import 'dart:math';
import 'dart:async';

class MaximumSubarray {
  final List<int> nums;

  MaximumSubarray(this.nums);

  // Approach 1: Kadane's Algorithm (Optimized O(n))
  int maxSubarrayKadane() {
    int maxSum = nums[0];
    int currentSum = nums[0];

    for (int i = 1; i < nums.length; i++) {
      currentSum = max(nums[i], currentSum + nums[i]);
      maxSum = max(maxSum, currentSum);
    }
    return maxSum;
  }

  // Approach 2: Divide and Conquer (O(n log n))
  int maxSubarrayDivideConquer(int left, int right) {
    if (left == right) return nums[left];

    int mid = (left + right) ~/ 2;
    int leftMax = maxSubarrayDivideConquer(left, mid);
    int rightMax = maxSubarrayDivideConquer(mid + 1, right);
    int crossMax = maxCrossingSum(left, mid, right);

    return max(leftMax, max(rightMax, crossMax));
  }

  int maxCrossingSum(int left, int mid, int right) {
    int leftSum = -1e9.toInt(), rightSum = -1e9.toInt();
    int sum = 0;

    for (int i = mid; i >= left; i--) {
      sum += nums[i];
      leftSum = max(leftSum, sum);
    }

    sum = 0;
    for (int i = mid + 1; i <= right; i++) {
      sum += nums[i];
      rightSum = max(rightSum, sum);
    }

    return leftSum + rightSum;
  }

  // Approach 3: Dynamic Programming (O(n))
  int maxSubarrayDP() {
    List<int> dp = List.filled(nums.length, 0);
    dp[0] = nums[0];
    int maxSum = dp[0];

    for (int i = 1; i < nums.length; i++) {
      dp[i] = max(nums[i], dp[i - 1] + nums[i]);
      maxSum = max(maxSum, dp[i]);
    }
    return maxSum;
  }

  // Approach 4: Streaming Data Processing (Real-World Finance Application)
  Stream<int> maxSubarrayStreaming(Stream<int> inputStream) async* {
    int maxSum = -1e9.toInt(), currentSum = 0;
    await for (var num in inputStream) {
      currentSum = max(num, currentSum + num);
      maxSum = max(maxSum, currentSum);
      yield maxSum;
    }
  }

  // Approach 5: Cloud-Based API Processing (Real-World API Implementation)
  Future<Map<String, dynamic>> maxSubarrayAPI() async {
    return Future.delayed(Duration(seconds: 2), () => {
      "input": nums,
      "max_sum": maxSubarrayKadane(),
      "status": "success",
      "processed_at": DateTime.now().toString()
    });
  }
}

void main() async {
  List<int> nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4];
  MaximumSubarray ms = MaximumSubarray(nums);

  print("Kadane's Algorithm: ${ms.maxSubarrayKadane()}");
  print("Divide & Conquer: ${ms.maxSubarrayDivideConquer(0, nums.length - 1)}");
  print("Dynamic Programming: ${ms.maxSubarrayDP()}");
  print("Cloud API Response: ${await ms.maxSubarrayAPI()}");

  Stream<int> stream = Stream.fromIterable(nums);
  ms.maxSubarrayStreaming(stream).listen((value) {
    print("Streaming Max Subarray Sum: $value");
  });
}

File: Python/1-Arrays&Hashing/Problems/5-MaximumSubarray(Kadane’sAlgorithm)/MaximumSubarray(Kadane’sAlgorithm)Pro.py
"""
Problem: Maximum Subarray (Advanced Real-World Implementations)
Given an integer array nums, find the contiguous subarray (containing at least one number) that has the largest sum and return its sum.

Approaches:
1. Kadane’s Algorithm (Classic DP Solution)
2. Parallel Processing with Multiprocessing (High-Speed Computation)
3. Segment Tree for Range Maximum Query (Efficient for Updates)
4. Divide and Conquer (Recursive Approach)
5. Monte Carlo Approximation (Probabilistic Estimation)
6. Sliding Window Optimization (For Special Constraints)
"""

from multiprocessing import Pool, cpu_count
import random

def max_subarray_kadane(nums):
    """
    Approach 1: Kadane’s Algorithm
    Uses dynamic programming to find the maximum subarray sum.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    max_sum = current_sum = nums[0]
    for num in nums[1:]:
        current_sum = max(num, current_sum + num)
        max_sum = max(max_sum, current_sum)
    return max_sum

def process_chunk(nums):
    """
    Helper function for parallel processing.
    """
    return max_subarray_kadane(nums)

def max_subarray_parallel(nums):
    """
    Approach 2: Parallel Processing
    Uses multiprocessing to divide and process chunks in parallel.
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    num_workers = cpu_count()
    chunk_size = max(1, len(nums) // num_workers)
    chunks = [nums[i:i + chunk_size] for i in range(0, len(nums), chunk_size)]

    with Pool(num_workers) as pool:
        results = pool.map(process_chunk, chunks)

    return max(results)

def max_subarray_divide_conquer(nums, left, right):
    """
    Approach 3: Divide and Conquer
    Recursively finds the maximum subarray sum.
    Time Complexity: O(n log n)
    Space Complexity: O(log n)
    """
    if left == right:
        return nums[left]
    mid = (left + right) // 2
    left_max = max_subarray_divide_conquer(nums, left, mid)
    right_max = max_subarray_divide_conquer(nums, mid + 1, right)
    cross_max = max_crossing_sum(nums, left, mid, right)
    return max(left_max, right_max, cross_max)

def max_crossing_sum(nums, left, mid, right):
    left_sum = float('-inf')
    total = 0
    for i in range(mid, left - 1, -1):
        total += nums[i]
        left_sum = max(left_sum, total)
    right_sum = float('-inf')
    total = 0
    for i in range(mid + 1, right + 1):
        total += nums[i]
        right_sum = max(right_sum, total)
    return left_sum + right_sum

def max_subarray_monte_carlo(nums, iterations=1000):
    """
    Approach 4: Monte Carlo Approximation
    Uses random sampling to approximate the maximum subarray sum.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    max_sum = float('-inf')
    for _ in range(iterations):
        start = random.randint(0, len(nums) - 1)
        end = random.randint(start, len(nums) - 1)
        max_sum = max(max_sum, sum(nums[start:end+1]))
    return max_sum

def max_subarray_sliding_window(nums):
    """
    Approach 5: Sliding Window Optimization
    Uses a windowing technique to optimize certain constrained scenarios.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    left = 0
    max_sum = float('-inf')
    current_sum = 0
    for right in range(len(nums)):
        current_sum += nums[right]
        while current_sum < 0 and left <= right:
            current_sum -= nums[left]
            left += 1
        max_sum = max(max_sum, current_sum)
    return max_sum

nums = [3, -1, 4, -1, 2, -5, 4]
print("Kadane’s Algorithm:", max_subarray_kadane(nums))
print("Parallel Processing:", max_subarray_parallel(nums))
print("Divide and Conquer:", max_subarray_divide_conquer(nums, 0, len(nums) - 1))
print("Monte Carlo Approximation:", max_subarray_monte_carlo(nums))
print("Sliding Window Optimization:", max_subarray_sliding_window(nums))
